File: chapter-1/linked_list.cpp
#include<iostream>
#include<bits/stdc++.h>

using namespace std;

struct Node{
    int value;
    struct Node *link;
};
int flag;

struct Node * insrt_Element(struct Node *head){
    struct Node *newNode,*ptr;
    int value;
    cout<<"enter the value: ";
    cin>>value;
    newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->value = value;
    if(head==NULL){
        newNode->link=NULL;
        head = newNode;
        cout<<"Linked List Created"<<endl;
    }
    else{
        ptr = head;
        while(ptr->link!=NULL){
            ptr= ptr->link;
        }
        ptr->link = newNode;
        newNode->link = NULL;
    }
    flag++;
    return head;
};

struct Node * dlt_Element(struct Node * head,int position){
    struct Node *p = head;
    struct Node *temp;

    if(p==NULL){
        cout<<"list is empty!! "<<endl;
    }
    else if(position==1){
        head = p->link;
        flag--;
    }
    else if(position>flag){
        cout<<"enough elements are not present"<<endl;
    }
    else{
        for(int i=1;i<position;i++){
            if(p!=NULL){
                temp = p;
                p = p->link;
            }
        }
        temp->link= temp->link->link;
        free(p);
        flag--;
    }
    return head;
};


void display(struct Node *head){
    struct Node *p=head;
    if(p==NULL){
        cout<<"empty list!"<<endl;
    }
    else{
         cout<<"Element are : ";
        while(p!=NULL){
            cout<<p->value<<" ";
            p = p->link;
        }
    }
    cout<<endl;
}


int main(){
    struct Node *head = NULL;
    int task;
    do{
        cout<<"1. Insert New Element"<<endl;
        cout<<"2. Display The List"<<endl;
        cout<<"3. Delete The element"<<endl;
        cout<<"0. For Exit"<<endl;
        cout<<"enter the task you want to perform : ";
        cin>>task;
        switch(task){
        case 0:
            break;
        case 1:
            head = insrt_Element(head);
            cout<<endl;
            break;

        case 2:
            display(head);
            cout<<endl;
            break;

        case 3:
            int num;

            cout<<"Enter the position where you delete the element:"<<endl;
            cin>>num;
            head = dlt_Element(head,num);
            cout<<endl;
            break;

        default:
            cout<<"enter correct choice :"<<endl;
            cout<<endl;
            break;
        }
    }while(task!=0);
    return 0;
}

File: chapter-1/Linked_List.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next_val = None

class Linked_List:
    def __init__(self):
        self.head_val = None

    def add(self, new_value):
        new_node = Node(new_value)
        new_node.next_val = self.head_val
        self.head_val = new_node

    def delete_Node(self,key):
        temp = self.head_val
        if(temp is not None):
            if(temp.value == key):
                self.head_val = temp.next_val
                temp = None
                return
        while(temp is not None):

            if(temp.value == key):
                break
            prev = temp
            temp = temp.next_val
        if(temp == None):
            return
        prev.next_val = temp.next_val
        temp = None

    def list_print(self):
        print_val = self.head_val
        while(print_val):
            print(f"{print_val.value}  ")
            print_val = print_val.next_val


list = Linked_List()

list.add(1)
list.add(2)
list.add(3)
list.add(4)
list.add(5)
print("Linked list created")
list.list_print()
list.delete_Node(4);
print("After deletion element 4")
list.list_print()
